### PR TITLE
Implement HTTP API for clock driven games

### PR DESCRIPTION
**Pull request by frostburn**
Initiate a HTTP API that uses shim sockets on server launch.
Move game listing code to GameServer.
Add a helper method for getting the current deal for puyo duel.

### CODE REVIEW

File: lib/server/index.js [hunk 1]
```diff
@@ -52,6 +52,25 @@ class GameServer {
     process.stdout.write(`Connection from ${socket.conn.remoteAddress}\n`);
   }
 
+  listGames() {
```
Comment (RauliL) [inline lib/server/index.js]: Why would you want method like this?

Comment (RauliL) [inline lib/server/index.js]: You can already iterate `server.games` where instance of `GameServer` is available.

File: lib/server/index.js [hunk 2]
```diff
@@ -52,6 +52,25 @@ class GameServer {
...
+          id,
+          mode: game.mode,
+          playerCount: game.playerCount,
+          maximumPlayerCount: game.maximumPlayerCount,
+        };
+      }),
+    }
+  }
+
+  viewGame(id) {
```
Comment (RauliL) [inline lib/server/index.js]: How does this view anything? Name of the method is very confusing.

File: lib/api/index.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    const engine = this.games[uuid].engine;
+    if (engine) {
+      engine.addEvent(data);
+      return { success: true };
+    }
+    return { success: false };
+  }
+
+  static install(app, server) {
+    const api = new this(server);
```
Comment (RauliL) [inline lib/api/index.js]: Why expose the `API` class at all when you could only export this as function?

File: server.js [hunk 4]
```diff
@@ -79,6 +81,10 @@ function launchServer(options) {
       } else {
         app.use(express.static(path.join(__dirname, 'public')));
       }
+
+      app.use(bodyParser.json());
+      const api = API.install(app, gameServer);
```
Comment (RauliL) [inline server.js]: This local variable is never used anywhere.

File: lib/server/index.js (hunk 1 above)
Comment (frostburn) [inline lib/server/index.js]: You need to list games both from the API (GET) and the socket from the browser. The iteration is the same in both cases so having it here is DRY.

File: lib/server/index.js (hunk 2 above)
Comment (frostburn) [inline lib/server/index.js]: Named after the API endpoint. The functionality is simple enough so I'll move it there verbatim.

File: lib/api/index.js (hunk 3 above)
Comment (frostburn) [inline lib/api/index.js]: Good point.

File: lib/server/index.js (hunk 1 above)
Comment (RauliL) [inline lib/server/index.js]: At least return an array and rename the method to something like `getCurrentlyRunningGames`.

File: server.js (hunk 4 above)
Comment (frostburn) [inline server.js]: I don't know how lifetimes works. If I don't store the thing in a variable will it just get garbage collected? I need it to stay in memory as it keeps track of all of the HTTP connections.

File: lib/server/index.js (hunk 1 above)
Comment (frostburn) [inline lib/server/index.js]: Why "currently running"? It returns every game on the server.

**Comment by frostburn** [comment]
Did the things.

**Pull request merged**